### ISSUE TITLE
RFC - Auth workflows as part of AWSMobileClient

### ISSUE THREAD

**Issue by rohandubal** [aws-amplify/aws-sdk-ios#1158]
## Feature Request

With the launch of [AWSMobileClient](https://aws-amplify.github.io/docs/ios/authentication), we simplified lot of authentication/ authorization workflows.  We now want to extend `AWSMobileClient` to support additional use-cases outlined below:

As a developer:

1. AWSMobileClient functionality:

    1. I want to enable SAML based authorization
    1. I want to add social buttons in drop-in UI that launches hosted UI for the selected social provider
    1. I want to use hosted UI from showSignIn API
    1. I want to allow my end user to sign out of all devices
    1. I want to enable device remember features
    1. I want to enable developer authenticated identities flow for authorization

1. Drop-in UI in AWSMobileClient functionality:

    1. I want to disable sign up flow
    1. I want to disable forgot password flow
    1. I want to add custom input fields on the registration screen
    1. I want to customize UI elements which include Button, Label, InputBox
    1. I want to use the drop-in UI without any navigation controller
    1. I want to enable localization
    1. I want to switch text orientation from right to left instead of left to right (culture/language semantics)
    1. I want to hide logo image / show sign in screen independently
    1. I want to change the properties of logo - size, scale, fit, etc.
    1. I want to just show the registration field directly
    1. I want to just use the change password flow
    1. I want a customized main screen

Let's use this issue to track the progress of this feature. This will help us address the following issues which we have already got feature requests for:

* [#1094](https://github.com/aws-amplify/aws-sdk-ios/pull/1094)
* [#917](https://github.com/aws-amplify/aws-sdk-ios/pull/917)
* [#815](https://github.com/aws-amplify/aws-sdk-ios/issues/815)
* [#931](https://github.com/aws-amplify/aws-sdk-ios/issues/931)
* [#1005](https://github.com/aws-amplify/aws-sdk-ios/issues/1005)
* [#731](https://github.com/aws-amplify/aws-sdk-ios/issues/731)
* [#942](https://github.com/aws-amplify/aws-sdk-ios/issues/942)

As part of this work we would also like to capture your feedback on this issue. Please let us know which of the above use cases align with your requirements, since that will help us prioritize the issues. For e.g. if you want to use the drop-in without navigation controller, please comment `2.5` since that is use-case which captures it: `I want to use the drop-in UI without any navigation controller`.

In addition, please comment with your ideal workflow for your app. Additionally, the following are starting point questions to help guide your responses:

1. Are you using Cognito UserPools(authentication) and Cognito IdentityPools(authorization) both in your app?
1. Are you using drop-in UI in your app?
1. Do you want more functionality in drop-in UI or more/ easy-to-use APIs in AWSMobileClient to better provide functionality in your app UI?
1. Do you want to use drop-in UI as an onboarding mechanism and then move to your own UI before production **_or_** use available customizations to match drop-in UI to your app theme?
1. Do you have social providers authentication in your app?
1. How many social providers (and which) are part of your app?
1. Do you prefer native social providers' SDK(e.g. FBLoginSDK) or hosted solutions(CognitoAuth)?
1. How important is framework size?
1. Do you have more than 1 configuration for same resources in your app? (e.g. 2 UserPools for the same app based on region)
1. Do you have SAML federation in your app?
1. Do you have features to remember user device in your app?

If you have a use-case which you want us to support in `AWSMobileClient` for authentication/ authorization use case, please comment below so that we can discuss and update the spec.

**Comment by BillBunting** [comment]
1 iv. _I want to allow my end user to sign out of all devices_. That would be nice.
1 v. _I want to enable device remember features_ That would be nice. 

2 iii. _I want to add custom input fields on the registration screen_
That would be helpful. Currently, my app has its own profile view to collect additional user information (like a public display name, Instagram account, and user bio).  I could add them to the sign-up workflow on the registration screen then store them in the user's profile. Now, after the user completes the initial AWSMobileClient sign up, then, I display a secondary profile view for them to complete (whenever there is missing required data) 

2 v. _I want to use the drop-in UI without any navigation controller_ 
Maybe, but since my app has a navigation controller, this is not an issue for my app. However, being able to move sign-up code into the AppDelegate or other locations where nav controller is not easily available would be nice. My app calls a "checkLogin" function that also obtains the identityId prior to making calls to AWS services; passing the nav controller around is a bit tedious which often includes doing so from a background task. Currently, there is no reliable "is logged in" method that I have had success with. (I'm waiting on some bug fixes, and hope they help) 

2 ix _I want to change the properties of logo - size, scale, fit, etc._
and xii _I want a customized main screen_
Absolutely, the more I can customize the logo and sign-in screen to match the app's branding, so it does not look like a "cookie cutter" app, the better. 

In addition, I would like direct support for Objective C. Currently, I wrap the Swift AWSMobileClient in a Swift AWSMobileClientHelper (singleton) class and make calls from Objective C through it (to abstract changes to AWSMobileClient, etc. from the remainder of the app written in Objective C, initially a MobileHub project in Objective C).

-------------
_In addition, please comment with your ideal workflow for your app. Additionally, the following are starting point questions to help guide your responses:_

_Are you using Cognito UserPools(authentication) and Cognito IdentityPools(authorization) both in your app?_ 
Yes, both. 

_Are you using drop-in UI in your app?_ Yes

_Do you want more functionality in drop-in UI or more/ easy-to-use APIs in AWSMobileClient to better provide functionality in your app UI?_ 
As a developer, I want to use AWSMobileClient to provide the sign-up, and sign-in functionality, so I can focus on the mission/functional code of my app without needing to develop and manage sign-in, sign-up code.

_Do you want to use drop-in UI as an onboarding mechanism and then move to your own UI before production or use available customizations to match drop-in UI to your app theme?_
I want to customize the drop-in UI with colors and a logo or other screen attributes to match the branding of the app (and not look exactly like all the other apps using the sign-un UI)

_Do you have social providers authentication in your app?_
Yes

_How many social providers (and which) are part of your app?_
Facebook and Google

_Do you prefer native social providers' SDK(e.g. FBLoginSDK) or hosted solutions(CognitoAuth)?_
No preference 

_How important is framework size?_
As long as it is reasonable. 

_Do you have more than 1 configuration for same resources in your app? (e.g. 2 UserPools for the same app based on region)_
No.

_Do you have SAML federation in your app?_
No.

_Do you have features to remember user device in your app?_
Not directly. (Using Pinpoint and use the identityId as a userId for saving endpoints for push notifications)

**Comment by dennispost** [comment]
_Are you using Cognito UserPools(authentication) and Cognito IdentityPools(authorization) both in your app?_ 
No, only Cognito UserPools

_Are you using drop-in UI in your app?_
No

_Do you want more functionality in drop-in UI or more/ easy-to-use APIs in AWSMobileClient to better provide functionality in your app UI?_
No drop-in UI, but easy-to-use APIs for login, logout, password change, mfa...

_Do you want to use drop-in UI as an onboarding mechanism and then move to your own UI before production or use available customizations to match drop-in UI to your app theme?_
I'm not using drop-in UI at all.

_Do you have social providers authentication in your app?_
Yes

_How many social providers (and which) are part of your app?_
1. Facebook

_Do you prefer native social providers' SDK(e.g. FBLoginSDK) or hosted solutions(CognitoAuth)?_
Hosted solutions

_How important is framework size?_
As long as it is reasonable. More important are built times. 

_Do you have more than 1 configuration for same resources in your app? (e.g. 2 UserPools for the same app based on region)_
No

_Do you have SAML federation in your app?_
No

_Do you have features to remember user device in your app?_
Yes

**Comment by ryancrunchi** [comment]
* AWSMobileClient functionality:

  I would like to use custom auth with AWSMobileClient and if possible without interactive auth. I found interactive auth is too much interferring on how we build our view controllers flows.


* Drop-in UI in AWSMobileClient functionality:

I don't like the fact that a business logic SDK (auth) also provides UI. I prefer keeping all control on my UI.

1. Are you using Cognito UserPools(authentication) and Cognito IdentityPools(authorization) both in your app?
Yes

2. Are you using drop-in UI in your app?
No

3. Do you want more functionality in drop-in UI or more/ easy-to-use APIs in AWSMobileClient to better provide functionality in your app UI?
No

4. Do you want to use drop-in UI as an onboarding mechanism and then move to your own UI before production **_or_** use available customizations to match drop-in UI to your app theme?
No

5. Do you have social providers authentication in your app?
Yes

6. How many social providers (and which) are part of your app?
Facebook

7. Do you prefer native social providers' SDK(e.g. FBLoginSDK) or hosted solutions(CognitoAuth)?
Prefer hosted centralized solutions if fully customizable.

8. How important is framework size?
It should not pass few Mb if it only contains source code. Build time is also important.

9. Do you have more than 1 configuration for same resources in your app? (e.g. 2 UserPools for the same app based on region)
No

10. Do you have SAML federation in your app?
No

11. Do you have features to remember user device in your app?
No

**Comment by motivus** [comment]
I'm confused about where to ask a question about customization of the Drop-In UI... but for my Cognito User Pool all I require of the user is their email address and password. Username and phone number are not required. Can someone tell me how to limit the sign up view to just those two elements? The "User Attributes" section of the [docs](https://aws-amplify.github.io/docs/ios/authentication) just seems to show hard-coding of certain parameters. How do I set which ones show on the UI?

**Comment by shannon-hager-skookum** [comment]
- Drop-in UI in AWSMobileClient functionality:
3. I want to add custom input fields on the registration screen
The signup fields in the drop-in UI did not match our needs the last time I checked and there was no way to fix this. Maybe it has changed?

5. I want to use the drop-in UI without any navigation controller
Navigation controller is needed during sign up but I normally prefer to present a modal during app use if login is needed.

6. I want to enable localization
7. I want to switch text orientation from right to left instead of left to right (culture/language semantics)
6 & 7 are not current needs but future needs.

1. Are you using Cognito UserPools(authentication) and Cognito IdentityPools(authorization) both in your app? 
Cognito UserPools only.

2. Are you using drop-in UI in your app? 
No. It seemed to work for a very specific signup use case and that didn't match our use case.

3. Do you want more functionality in drop-in UI or more/ easy-to-use APIs in AWSMobileClient to better provide functionality in your app UI? 
I would rather see the effort put toward the API because I can't use the drop-in UI. If you added functionality to the UI to meet my use cases, there would be others who still couldn't use it but all of us would benefit from API updates.

4. Do you want to use drop-in UI as an onboarding mechanism and then move to your own UI before production **_or_** use available customizations to match drop-in UI to your app theme?
If I used it, I would want it to match my UI. Otherwise it looks unprofessional.

5. Do you have social providers authentication in your app?
No.
6. How many social providers (and which) are part of your app?
None.

7. Do you prefer native social providers' SDK(e.g. FBLoginSDK) or hosted solutions(CognitoAuth)?
N/A

8. How important is framework size?
Medium.

9. Do you have more than 1 configuration for same resources in your app? (e.g. 2 UserPools for the same app based on region)
Not yet but eventually.

10. Do you have SAML federation in your app?
No.

11. Do you have features to remember user device in your app?
No.

**Comment by shannon-hager-skookum** [comment]
Update after giving the drop-in UI another shot:

**Requiring a navigation controller** isn't a deal-breaker but it's close.

Not being able to **add/remove the fields on the signup form** is a deal-breaker. I'm guessing it is a deal-breaker for most cases and an inconvenience for most others.

Not having checkboxes and links to **Terms of Service and Privacy Policy on the signup** for is a deal-breaker.
All new iOS apps require a privacy policy link. I can make a workaround signup flow that has a separate privacy policy view before or after signing up but that's not something I would do.

**Field validation** as it currently exists is a dealbreaker.
Password requirements and the expected phone number format is a shot in the dark and validated in random order. Putting in "a" as the password and "310.555.1212" as the phone number results in this alert:
![Screen Shot 2019-03-11 at 11 02 07 AM](https://user-images.githubusercontent.com/18582836/54133911-53938e00-43ed-11e9-9489-3ec571e4ac1c.png)
Is that supposed to be an ellipsis at the end or is it requiring 2 dots? It's confusing. 
I add "+1'' to the phone number that so now it is "+1310.555.1212" then submit again and get this error:
![Screen Shot 2019-03-11 at 10 55 06 AM](https://user-images.githubusercontent.com/18582836/54133525-9acd4f00-43ec-11e9-8ddc-10e9abb66e8c.png)
(Yes, that is the actual message shown to the user.)
But no more messages about the phone number so I guess the phone is good now.
I add a few more characters to the password so it is 6 chars long and submit again. The "InvalidParameterException" message above says "Member must have length greater than or equal to 6" but my 6-char password ("asdfgh") results in this:
![Screen Shot 2019-03-11 at 11 04 49 AM](https://user-images.githubusercontent.com/18582836/54134125-bb49d900-43ed-11e9-924e-52dd1197ab30.png)
So I make a 7-char password ("asdfghj") and get the same error:
![Screen Shot 2019-03-11 at 11 04 49 AM](https://user-images.githubusercontent.com/18582836/54134125-bb49d900-43ed-11e9-924e-52dd1197ab30.png)
So I make an 8-char password ("asdfghjk") and get a new error: 
![Screen Shot 2019-03-11 at 11 14 10 AM](https://user-images.githubusercontent.com/18582836/54134702-d2d59180-43ee-11e9-9a28-a1173ae0d81e.png)
So I change my password to "Asdfghjk" and get this error:
![Screen Shot 2019-03-11 at 11 15 46 AM](https://user-images.githubusercontent.com/18582836/54134839-0adcd480-43ef-11e9-8d1e-39acd0189dcf.png)
So I change my password to "Asdfghj1" and get this error:
![Screen Shot 2019-03-11 at 11 16 29 AM](https://user-images.githubusercontent.com/18582836/54134892-2647df80-43ef-11e9-96e9-aee0fabba516.png)
So I change my password to "Asdfgh1!" and get this error:
![Screen Shot 2019-03-11 at 11 22 01 AM](https://user-images.githubusercontent.com/18582836/54135291-e9301d00-43ef-11e9-905e-2d5a6eb65014.png)
No mention of the correct format, the only phone number format mentioned in any of the messages is `+1992..` but that isn't helpful in this case.
I replace the dots in my phone number with dashes, changing it to "+1310-555-1212" and get the same error.
I change the phone number to the format that is as close to universally accepted as any I can know of: "+1 (310) 555-1212" and get the same error.
I change the phone number to "+13105551212" and it works.

(Lego Batman voice: "First try!")

In summary:
- Fields validation requirements are checked in random order
- Validation is broken and/or the messages are incorrect
- The validation titles e.g. "InvalidPasswordException" are not acceptable for user-facing alert UI.
- The validation message content is similarly unsuitable in that it refers to "member" instead of "password", displays regex, etc.
- On iOS, the "OK" button should be spelled "OK" not "Ok".
- No way to tell the user the requirements before they fill in the form.
- No way to tell the user the requirements after they submit a form that fails validation, forcing them to trial-and-error their way through.

Maybe all of the above can be customized and I'm missing it, I don't see how there is a question of whether the drop-in signin/signup flow is good enough for use except in the case of internal proof-of-concept apps or apps that require sign-up elsewhere (on a web site, company ex/intranet, etc).

If this level of detail is too much for this thread, let me know where it should go and I'll create a new issue.

**Comment by kvasukib** [comment]
Request for customizing SignUp UI (https://github.com/aws-amplify/aws-sdk-ios/issues/1277)

**Comment by wallowabk** [comment]
Here's an auth flow use case to consider:

[Password-free SMS Authentication with AWS Cognito, Lambda Node.js & iOS Swift](https://quanta.svbtle.com/mobile-number-only-authentication-with-aws-cognito-ios-swift)

**Comment by Gringox** [comment]
Hi team,

Is support for custom auth in AWSMobileClient on the roadmap?

Thanks.

**Comment by desokroshan** [comment]
Add support for user migration, requested in https://github.com/aws-amplify/aws-sdk-ios/issues/1651

**Comment by palpatim** [comment]
Additional customer request:

When an automatic token refresh encounters a situation where the user must sign in (e.g., the refresh token has expired), if there is no UI delegate assigned [sic--it's unclear exactly what is meant by this or how we would detect it], do not wait and ask for a sign-in flow, but rather return an error. This could be accomplished by a configuration option to suppress the "sign in wait" behavior.

**Comment by mcroker** [comment]
Are you using Cognito UserPools(authentication) and Cognito IdentityPools(authorization) both in your app?
>>>Both

Are you using drop-in UI in your app?
>>> No

Do you want more functionality in drop-in UI or more/ easy-to-use APIs in AWSMobileClient to better provide functionality in your app UI?
>>> More easy to use APIs.

Do you want to use drop-in UI as an onboarding mechanism and then move to your own UI before production or use available customizations to match drop-in UI to your app theme?
>>> Not using drop-in UI

Do you have social providers authentication in your app?
>>> Yes

How many social providers (and which) are part of your app?
>>> Google, Facebook, Apple

Do you prefer native social providers' SDK(e.g. FBLoginSDK) or hosted solutions(CognitoAuth)?
>>> For mobile app prefer native (especially for SignInWithApple which is a much better experience native).  But ideally would be able to use same social providers to authenticate from Web. 

How important is framework size?

Do you have more than 1 configuration for same resources in your app? (e.g. 2 UserPools for the same app based on region)
>>>No

Do you have SAML federation in your app?
>>> No

Do you have features to remember user device in your app?
>>> No (other than that which come via anon IDP user)

**Comment by panindravarma** [comment]
Hi team,

Is there any update on support for Cognito User Pool custom auth in AWSMobileClient?

Thanks.

**Comment by palpatim** [comment]
@panindravarma 

> Is there any update on support for Cognito User Pool custom auth in AWSMobileClient?

AWSMobileClient [already supports User Pools custom auth](https://docs.amplify.aws/sdk/auth/custom-auth-flow/q/platform/ios).

Hope this helps.

**Comment by dakrawczyk** [comment]
Hi all, this is the closest thread I could find to password policies. I am wondering if it is possible to access the `AWSModel` object `AWSCognitoIdentityProviderPasswordPolicyType` (found here: [link](https://github.com/aws-amplify/aws-sdk-ios/blob/6dfaf175a3b551ea3b1f8764f71bceda8942a2f1/AWSCognitoIdentityProvider/AWSCognitoIdentityProviderModel.m#L4163))

We are trying to guide the user to create a valid password in our UI (based on whatever the password policy we define) without needing to show the AWS Supplied error message, which isn't very user friendly: 
> validation error detected: Value at 'proposedPassword' failed to satisfy constraint: Member must have length greater than or equal to 6

**Comment by atierian** [comment]
Thanks for all of the feedback here. We're including these in the prioritization of work on [Amplify Swift](https://docs.amplify.aws/swift/).
